### ISSUE TITLE
isinstance(..., tuple) causes "partially unknown" error

### ISSUE THREAD

**Issue by catwell** [microsoft/pyright#9025]
**Describe the bug**

Testing if a variable is a tuple appends `tuple[Unknown, ...]` to its type and causes "is partially unknown" even when it should not.

This is a regression since 1.1.380 (no error in 1.1.379, still present in 1.1.381).

**Code or Screenshots**

This obviously works:

```py
from typing import Any


def f() -> Any:
    result: Any = None
    return result
```

but this equivalent code causes an error:

```py
from typing import Any


def f() -> Any:
    result: Any = None
    if isinstance(result, tuple):
        pass
    return result
```

```
foo.py:8:12 - error: Return type, "tuple[Unknown, ...] | Any", is partially unknown (reportUnknownVariableType)
```

**Comment by erictraut** [comment]
Pyright is working as intended here, so I don't consider this a bug.

The `reportUnknownVariableType` check is off by default. I presume that you've either manually enabled it or are using "strict" type checking mode, which also enables this check. When this is enabled, pyright tells you when you're using a type that includes an `Unknown` (an "implicit `Any`). 

If you don't want pyright to report issues with unknown or partially-unknown types, you can leave this check disabled.

In this case, the `isinstance` type narrowing results in a value of type `tuple[Unknown, ...]`, which triggers the check.

Version 1.1.380 includes a bunch of bug fixes in the `isinstance` type narrowing code path, which is why you are seeing a recent change in behavior here.

**Comment by catwell** [comment]
@erictraut Yes, this is in strict mode. I understand what happens, but I do not think it is very logical. More precisely I do not understand why the return type is narrowed by the addition of the `isinstance` test. Should it not be be narrowed inside the `if` block but not after it?

**Comment by erictraut** [comment]
Narrowed types propagate beyond guarded blocks. When an `if` or `else` block terminates without a control flow statement (like `raise`, `return`, or `continue`), control "merges" back with the code outside of the block, and all types that were narrowed within the `if` block are likewise merged (via union).

This may conceptually make more sense if you think of your code as a directed code flow graph. If narrowed types didn't propagate beyond guarded blocks, it would cause many problems. Consider the following example:

```python
def func(x: object) -> int:
    if isinstance(x, int):
        print("already an int!")
    else:
        x = 0

    return x
```

**Comment by catwell** [comment]
Yeah, it makes sense... I suppose the weird thing actually comes from the fact that `tuple[Unknown, ...]` is not a subtype of `Any`.

Anyway thanks, works as expected.
